subgraphs/collectible-market/mappings/index.ts

```diff
@@ -0,0 +1,263 @@
+  collection.save();
+}
+
+/**
+ * ASK ORDERS
+ */
+
+export function handleAskNew(event: AskNew): void {
+  // 1. User
+  let user = User.load(event.params.seller.toHex());
+  if (user == null) {
+    user = new User(event.params.seller.toHex());
+    user.numberTokensListed = ONE_BI;
+    user.numberTokensPurchased = ZERO_BI;
+    user.numberTokensSold = ZERO_BI;
+    user.totalVolumeInBNBTokensPurchased = ZERO_BD;
+    user.totalVolumeInBNBTokensSold = ZERO_BD;
+    user.totalFeesCollectedInBNB = ZERO_BD;
+    user.averageTokenPriceInBNBPurchased = ZERO_BD;
+    user.averageTokenPriceInBNBSold = ZERO_BD;
+  } else {
+    user.numberTokensListed = user.numberTokensListed.plus(ONE_BI);
+  }
+
+  // 2. Collection
+  let collection = Collection.load(event.params.collection.toHex());
+  collection.numberTokensListed = collection.numberTokensListed.plus(ONE_BI);
+
+  // 3. Token
+  let tokenConcatId = event.params.collection.toString() + "-" + event.params.tokenId.toString();
+  let token = NFT.load(tokenConcatId);
+
+  if (token == null) {
+    token = new NFT(tokenConcatId);
+    token.tokenId = event.params.tokenId;
+    token.collection = event.params.collection.toHex();
+    token.metadataUrl = fetchTokenURI(event.params.collection, event.params.tokenId);
+
+    token.latestTradedPriceInBNB = ZERO_BD;
+    token.tradeVolumeBNB = ZERO_BD;
+    token.totalTrades = ZERO_BI;
+
+    token.isTradable = true;
+  }
+
+  token.isTradable = true;
+
+  user.save();
+  token.save();
+  collection.save();
+}
+
+export function handleAskCancel(event: AskCancel): void {
+  let user = User.load(event.params.seller.toHex());
+  user.numberTokensListed = user.numberTokensListed;
+
+  let collection = Collection.load(event.params.collection.toHex());
+  collection.numberTokensListed = collection.numberTokensListed.minus(ONE_BI);
+
+  let tokenConcatId = event.params.collection.toString() + "-" + event.params.tokenId.toString();
+  let token = NFT.load(tokenConcatId);
+
+  token.isTradable = true;
+
+  user.save();
+  collection.save();
+  token.save();
+}
+
+/**
+ * BUY ORDERS
+ */
+
+export function handleTrade(block: ethereum.Block, event: Trade): void {
+  // 1. Transaction
+  let transactionConcId =
+    event.params.collection.toString() +
+    "-" +
+    event.params.tokenId.toString() +
+    "-" +
+    event.transaction.hash.toHexString();
+
+  let transaction = new Transaction(transactionConcId);
+  transaction.block = block.number;
+  transaction.timestamp = block.timestamp;
+  transaction.collection = event.params.collection.toHex();
+  transaction.tokenId = event.params.collection.toString() + "-" + event.params.tokenId.toString();
+  transaction.askPrice = event.params.price.toBigDecimal().div(EIGHTEEN_BD);
+  transaction.netPrice = event.params.netPrice.toBigDecimal().div(EIGHTEEN_BD);
+  transaction.buyer = event.params.buyer.toHex();
+  transaction.seller = event.params.seller.toHex();
```

Comment:
Your subgraph will crash if either the `buyer` or `seller` doesn't exists in `User` entity collection. Let's put this transaction mapping at the end and ensure both entities exists first. And please, reference them as `buyer.id` and `seller.id`.

Reply:
Adjusted